File: 1.arrays.and.strings/1.2.valid.palindrome.cpp
/**
 * @file 1.2.valid.palindrome
 * @author Touhid Alam <taz.touhid@gmail.com>
 *
 * @date Monday December 14 2020
 *
 * @brief 
 */

#include <bits/stdc++.h>

using namespace std;

class Solution {
 public:
  // O(n)
  bool isPalindrome(string s) {
    int n = s.size(), l = 0, r = n - 1;

    while (l < r) {
      while (l < r && !isalnum(s[l])) ++l;
      while (l < r && !isalnum(s[r])) --r;
      if (tolower(s[l++]) != tolower(s[r--])) return 0;
    }
    return 1;
  }
};

File: 1.arrays.and.strings/1.2.valid.palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int n = s.length();

        for (int l = 0, r = n - 1; l < r;) {
            if (!Character.isLetterOrDigit(s.charAt(l))) {
                ++l;
                continue;
            } else if (!Character.isLetterOrDigit(s.charAt(r))) {
                --r;
                continue;
            } else if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r)))
                return false;
            else {
                ++l;
                --r;
            }
        }
        return true;
    }
}